app/javascript/src/hyacinth_v1/components/projects/publish_targets/PublishTargetsForm.js

```diff
@@ -0,0 +1,158 @@
+import React, { useEffect, useState } from 'react';
+import { useMutation, useQuery } from '@apollo/react-hooks';
+import PropTypes from 'prop-types';
+import {
+  Row, Col, Form, Badge,
+} from 'react-bootstrap';
+import { useHistory } from 'react-router-dom';
+import produce from 'immer';
+import GraphQLErrors from '../../shared/GraphQLErrors';
+import FormButtons from '../../shared/forms/FormButtons';
+import { getProjectsPublishTargetsQuery, updateProjectsPublishTargetsMutation } from '../../../graphql/projects/publishTargets';
+
+const PublishTarget = (props) => {
+  const {
+    stringKey, enabled, readOnly, enabledDataCallback,
+  } = props;
+  const onEnable = (event) => {
+    const { target: { checked } } = event;
+    if (enabled !== checked) {
+      enabledDataCallback(stringKey, { enabled: checked });
+    }
+  };
+
+  return (
+    <Row key={stringKey} id={stringKey} className="mb-2" style={{ backgroundColor: '#eaeaea', border: 'none' }}>
+      <Col md={2}>
+        <Badge bg="info">{stringKey}</Badge>
+      </Col>
+      <Col xs={4} md={2}>
+        <Form.Check
+          id={`enabled_${stringKey}`}
+          type="checkbox"
+          checked={enabled ? 'checked' : ''}
+          label="Enabled"
+          name="enabled"
+          onChange={onEnable}
+          className="align-middle"
+          inline
+          disabled={readOnly}
+        />
+      </Col>
+    </Row>
+  );
+};
+
+const mapPublishTargetData = (projectsPublishTargetsData) => {
+  const publishTargetData = {};
+
+  projectsPublishTargetsData.projectsPublishTargets.forEach((publishTarget) => {
+    const { stringKey, enabled } = publishTarget;
+
+    publishTargetData[stringKey] = { enabled };
+  });
+  return publishTargetData;
+};
+
+export const PublishTargetsForm = ({ readOnly, projectStringKey }) => {
+  const history = useHistory();
+
+  const [projectPublishTargets, setProjectPublishTargets] = useState({});
+
+  const variables = { stringKey: projectStringKey };
+
+  const {
+    loading: projectsPublishTargetsLoading,
+    error: projectsPublishTargetsError,
+    data: projectsPublishTargetsData,
+  } = useQuery(getProjectsPublishTargetsQuery, { variables });
+
+  useEffect(() => {
+    if (projectsPublishTargetsLoading === false && projectsPublishTargetsData) {
+      const mappedQueryData = mapPublishTargetData(projectsPublishTargetsData);
+      setProjectPublishTargets(mappedQueryData);
+    }
+  }, [projectsPublishTargetsLoading, projectsPublishTargetsData]);
+
+  const [updateEnabledPublishTargets, { error: updateError }] = useMutation(updateProjectsPublishTargetsMutation);
+
+  if (projectsPublishTargetsLoading) return (<></>);
```

Comment:
If you move `if (projectsPublishTargetsLoading) return (<></>);` above your useEffect, will that simplify things?  Then you wouldn't need to check if `projectsPublishTargetsLoading === false && projectsPublishTargetsData` inside the useEffect function.

Reply:
I was working around an infinite rerender from the state modification using the pattern you describe, but maybe I missed something?

Reply:
Moving the `availablePublishTargets` query to a field on Project allowed me to roll the query up in the calling component, which removes the necessity of using state *and* query *and* mutation here, so the effect could be removed.